### ISSUE TITLE
with secret runs inside a false conditional

### ISSUE THREAD

**Issue by mbrancato** [hashicorp/consul-template#1357]
### Consul Template version

```
consul-template v0.24.0 (36bc20c9)
```

This affects Vault Agent as well.

### Configuration

```hcl
pid_file = "./pidfile"

vault {
   address = "http://127.0.0.1:8200"
  renew_token = false

}

template {
  source      = "./testing.tmpl"
  destination = "./testing.txt"
}
```

This config works as expected - the `with secret` Vault POST/PUT does not occur.
```liquid
{{- $testExists := 0 -}}
{{- range secrets "secret/metadata/" -}}
{{- if eq . "testing" -}}
{{- $testExists = 1 -}}
{{- end -}}
{{- end }}

result {{ $testExists }}
{{ if eq 1 0 }}
Secret Missing
{{- with secret "secret/data/testing" "value=x" }}
created
{{ . }}
{{- end }}
{{- end }}
```

This version uses a variable in the version and it does not honor the conditional and performs a write to the secret path.
```liquid
{{- $testExists := 0 -}}
{{- range secrets "secret/metadata/" -}}
{{- if eq . "testing" -}}
{{- $testExists = 1 -}}
{{- end -}}
{{- end }}

result {{ $testExists }}
{{ if eq $testExists 0 }}
Secret Missing
{{- with secret "secret/data/testing" "value=x" }}
created
{{ . }}
{{- end }}
{{- end }}```

```liquid
# Include sample data you reference in the template from Consul or Vault here.
```

### Command

Subsequent calls should not increment the secret version
```shell
$ consul-template -config=ct-config.hcl -dry -once
> ./testing.txt


result 1

map[data:map[value:x] metadata:map[created_time:2020-03-17T15:02:57.456389921Z deletion_time: destroyed:false version:21]]


$ consul-template -config=ct-config.hcl -dry -once
> ./testing.txt


result 1

map[data:map[value:x] metadata:map[created_time:2020-03-17T15:02:59.62248027Z deletion_time: destroyed:false version:22]]
```

### Debug output

https://gist.github.com/mbrancato/f187161d208eb59ee98be9c1432923b1

### Expected behavior

The `PUT` should only occur if the conditional evaluates as true.

### Actual behavior

The `PUT` happens when the conditional evaluates as false.

### Steps to reproduce

1. `vault kv put secret/testing value=y`
2. `consul-template -config=ct-config.hcl -dry -once`
3. `consul-template -config=ct-config.hcl -dry -once`

**Comment by eikenb** [comment]
Hey @mbrancato, thanks for the report.

Looks like you're running into [multi-phase-execution](https://github.com/hashicorp/consul-template#multi-phase-execution). 

When evaluating the template it makes multiple passes finding the dependencies each pass. So what is happening here is that on the first pass `{{- $testExists := 0 -}}` is evaluated and set, but `{{- range secrets "secret/metadata/" -}}` is registered as something that needs to be looked up. So `{{- $testExists = 1 -}}` doesn't get run the first time through. Thus when you get to `{{ if eq $testExists 0 }}`, it will always evaluate to true that first pass and so it will register the `{{- with secret "secret/data/testing" "value=x" }}` to be run.

On the latter passes it sets `$testExists` to 1 which it then sees fails the `if eq` test and so doesn't display anything in the `if` block, ignoring any text that would have been displayed in it.

TLDR; `with secret` is getting run every time even though that part of the template isn't displayed.

Make sense?

To work around multi-phase-evaluation you'll need the base/0 case (first pass) to do nothing and only to do something on later passes. I'm still thinking about the best way to do this and will post if/when I have something.. but thought I'd explain what is going on first in case you have an idea.

**Comment by mbrancato** [comment]
Thanks @eikenb - I've come up with a **partial** solution. Its not entirely stable as I've seen the version creep up as high as version 3 with a short ttl and re-rendering.

```liquid
{{- $testExists := 0 -}}
{{- $testPath := "false" -}}
{{- range secrets "secret/metadata/" -}}
{{- if eq . "testing" -}}
{{- $testPath = "testing" -}}
{{- $testExists = 1 -}}
{{- else -}}
{{- $testPath = "testing" -}}
{{- end -}}
{{- end }}

result {{ $testExists }}
{{ if eq $testExists 0 }}
Secret Missing
{{- secret (print "secret/data/" $testPath) "value=x" }}
{{- else -}}
{{- secret (print "secret/data/" $testPath) }}
{{- end }}
```

**Comment by eikenb** [comment]
I had something similar but didn't post it as it wouldn't handle the case of `{{- range secrets "secret/metadata/" -}}` having no values.

I don't think setting an initial value via a template is a common pattern as I can't come up with anything that I would consider 'good enough' given the existing tools. You might be better off seeing if you can initialize the secret via another means. Though if you stick with this and come up with a good solution, please post it (to help any others in a similar situation).

Thanks.
